File: second course/PDOO/civitasJava/src/civitas/Casilla.java
package civitas;

import java.util.ArrayList;

public class Casilla {
  
  private static int carcel = -1;
  
  private float importe;
  private String nombre;
  
  //Atributos de referencia
  private TipoCasilla tipo;
  private TituloPropiedad tituloPropiedad;
  private Sorpresa sorpresa;
  private MazoSorpresas mazo;
  
  
  private void init(){ //TESTED
		importe = 0f;
		nombre = null;
		tipo = null;
		tituloPropiedad = null;
		sorpresa = null;
		mazo = null;
  }
  
  Casilla(String nombre){ //TESTED
		init();
	
		switch(nombre){
			case "Salida":
				tipo = TipoCasilla.DESCANSO;
				break;
		
			case "Parking":
				tipo = TipoCasilla.DESCANSO;
				break;
		
			case "Carcel":
				tipo = TipoCasilla.DESCANSO;
		}
		this.nombre = nombre;
  }
  
  Casilla(TituloPropiedad titulo){ //TESTED
		init();
		tipo = TipoCasilla.CALLE;
		tituloPropiedad = titulo;
		nombre = titulo.getNombre();
  }
  
  Casilla(float cantidad, String nombre){ //TESTED
		init();
		tipo = TipoCasilla.IMPUESTO;
		this.nombre = nombre;
		importe = cantidad;
  }
  
  Casilla(int numCasillaCarcel, String nombre){ //TESTED
		init();
		tipo = TipoCasilla.JUEZ;
		this.nombre = nombre;
		carcel = numCasillaCarcel;
  }
  
  Casilla(MazoSorpresas mazo, String nombre){ //TESTED
		init();
		tipo = TipoCasilla.SORPRESA;
		this.nombre = nombre;
		this.mazo = mazo;
  }
  
  private void informe(int iactual, ArrayList<Jugador> todos){ //TESTED
		Diario.getInstance().ocurreEvento("El jugador " + todos.get(iactual).getNombre() + " ha caido en:\n" + toString());
  }
  
  private void recibeJugador_impuesto(int iactual, ArrayList<Jugador> todos){ //TESTED
		if(jugadorCorrecto(iactual, todos)){
			informe(iactual, todos);
			todos.get(iactual).pagaImpuesto(importe);
		}
  }
  
  private void recibeJugador_juez(int iactual, ArrayList<Jugador> todos){ //TESTED
		if(jugadorCorrecto(iactual, todos)){
			informe(iactual, todos);
			todos.get(iactual).encarcelar(carcel);
		}
  }
  
  @Override
  public String toString(){ //TESTED
		String info = "Casilla :" + nombre
                + "\n\timporte: " + importe
                + "\n\ttipo:" + tipo
                + "\n\tTitulo de propiedad: " + tituloPropiedad;
        
    return info;
  }
  
  public String getNombre(){ //TESTED
		return nombre;
  }
  
  TituloPropiedad getTituloPropiedad(){ //TESTED
		return tituloPropiedad;
  }
  
  public boolean jugadorCorrecto(int iactual, ArrayList<Jugador> todos){ //TESTED
		return (iactual >= 0 && iactual < todos.size());
  }
  
  private void recibeJugador_calle(int iactual, ArrayList<Jugador> todos){ //TESTED
		if(jugadorCorrecto(iactual, todos)){
			informe(iactual, todos);
	  
			Jugador jugador = todos.get(iactual);
	  
	  if(!tituloPropiedad.tienePropietario())
			jugador.puedeComprarCasilla();
	  else
			tituloPropiedad.tramitarAlquiler(jugador);
	  
		}
  }
  
  void recibeJugador(int iactual, ArrayList<Jugador> todos){ //TESTED
		switch(tipo){
      case CALLE:
        recibeJugador_calle(iactual,todos);
        break;
      case IMPUESTO:
        recibeJugador_impuesto(iactual,todos);
        break;
      case JUEZ:
        recibeJugador_juez(iactual,todos);
        break;
      case SORPRESA:
        recibeJugador_sorpresa(iactual,todos);
        break;
      default:
        informe(iactual,todos);                 
    }
  }
  
  private void recibeJugador_sorpresa(int iactual, ArrayList<Jugador> todos){ //TESTED
		if(jugadorCorrecto(iactual, todos)){
			informe(iactual, todos);
			sorpresa = mazo.siguiente();
			sorpresa.aplicarAJugador(iactual, todos);
		}
  }
  
  ///////////////////////////////////////////
   
  public static void main(String [] args){
		
		
		ArrayList<Jugador> jugadores = new ArrayList<>();
		jugadores.add(new Jugador("Juan"));
		jugadores.add(new Jugador("Paco"));
		
		MazoSorpresas m = new MazoSorpresas();
		Sorpresa s = new Sorpresa(TipoSorpresa.SALIRCARCEL, m);
		m.alMazo(s);
		
		Casilla c = new Casilla(m, "Sorpresa");
		
		c.recibeJugador_sorpresa(1, jugadores);
		
		while(Diario.getInstance().eventosPendientes())
			System.out.println(Diario.getInstance().leerEvento());
		
		System.out.println(jugadores.get(1).toString());

  }
}

File: second course/PDOO/civitasRuby/pkg/civitasRuby-0.0.1/lib/ casilla.rb
# encoding:utf-8

require_relative 'jugador'
require_relative 'tipo_casilla'
require_relative 'titulo_propiedad'
require_relative 'tipo_casilla'
require_relative 'sorpresa'

module Civitas
	class Casilla
		
		private_class_method :new
		
		attr_reader :nombre, :titulo_propiedad
		
		@@carcel = -1
		
		
		def initialize(importe, nombre, tipo, titulo_propiedad, sorpresa, mazo)
			@importe = importe
			@nombre = nombre
			@tipo = tipo
			@titulo_propiedad = titulo_propiedad
			@sorpresa = sorpresa
			@mazo = mazo
		end
		
		def self.new_por_nombre(nombre)
			new(0.0, nombre, Civitas::TipoCasilla::DESCANSO, nil, nil,nil)
		end
		
		def self.new_por_titulo_propiedad(titulo_propiedad)
			new(0.0, titulo_propiedad.nombre(), Civitas::TipoCasilla::CALLE, titulo_propiedad, nil, nil)
		end 
		
		def self.new_por_impuesto(cantidad, nombre)
			new(cantidad, nombre, Civitas::TipoCasilla::IMPUESTO, nil, nil, nil)
		end
		
		def self.new_por_juez(num_casilla_carcel, nombre)
			new(0.0, nombre, Civitas::TipoCasilla::JUEZ, nil, nil, nil)
			@@carcel = num_casilla_carcel
		end
		
		def self.new_por_mazo(mazo, nombre)
			new(0.0, nombre, Civitas::TipoCasilla::SORPRESA, nil, nil, mazo)
		end
		
		private
		
		def informe(iactual, todos)
			Diario.instance.ocurre_evento("El jugador " + todos[iactual].nombre() + "ha caido en \n" + to_string())
		end
		
		def recibe_jugador_impuesto(iactual, todos)
			if(jugador_correcto(iactual, todos))
				informe(iactual, todos)
				todos[iactual].paga_impuesto(@importe)
			end
		end
		
		def recibe_jugador_juez(iactual, todos)
			if(jugador_correcto(iactual, todos))
				informe(iactual, todos)
				todos[iactual].encarcelar(@carcel)
			end
		end
	
	#### PENDIENTES DE IMPLEMENTACION ####	
		
		def recibe_jugador_calle(iactual, todos)
			
		end
		
		def recibe_jugador(iactual, todos)
			
		end
		
		def recibe_jugador_sorpresa(iactual, todos)
			
		end
		
	########################################

		public
		
		def to_string()
			info = "Casilla : #{@nombre} 
											\n\timporte: #{@importe} 
											\n\tMazo: #{@mazo} 
											\n\tSorpresa: #{@sorpresa} 
											\n\ttipo: #{@tipo} 
											\n\tTitulo de propiedad: #{@titulo_propiedad} 
											\n\tCarcel: #{@carcel}"
			
			return info
		end
		
		def jugador_correcto(iactual, todos)
			return (iactual >= 0 && iactual < todos.size())
		end
		
	end
	
	casilla = Casilla.new_por_nombre("CR7")
	
	puts casilla.to_string();
	
end
